extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -313,74 +336,69 @@ public Operator.Signal push(RowsAndColumns rac)
       public void completed()
       {
         try {
-          // resultRowsAndCols has reference to frameRowsAndCols
-          // due to the chain of calls across the ops
-          // so we can clear after writing to output
-          flushAllRowsAndCols(resultRowAndCols);
-          frameRowsAndCols.clear();
-
+          flushAllRowsAndCols();
         }
         catch (IOException e) {
           throw new RuntimeException(e);
         }
-        finally {
-          frameRowsAndCols.clear();
-          resultRowAndCols.clear();
-        }
       }
     });
   }
 
   /**
-   * @param resultRowAndCols Flush the list of {@link RowsAndColumns} to a frame
+   * Flushes {@link #resultRowAndCols} to the frame starting from {@link #rowId}, upto the frame writer's capacity.
    * @throws IOException
    */
-  private void flushAllRowsAndCols(ArrayList<RowsAndColumns> resultRowAndCols) throws IOException
+  private void flushAllRowsAndCols() throws IOException
   {
     RowsAndColumns rac = new ConcatRowsAndColumns(resultRowAndCols);
```

Comment:
its just performance: but the creation of this rac is an `O(n)` operation; regardless where the `rowId` stands.
that's why it would have been better to just pack all these things into an inner-workhorse class....when that will be done this should be taken into account.